#### Contents

#### Checking for Mixed type Data

#### Check for and Addressing Missing Values

#### Finding and Address Duplicates

#### Task 4.5

# Importing Libraries

In [3]:
# Import libraries
import pandas as pd
import numpy as np
import os

## Importing Data Frames

In [4]:
# Making Path do data as a string
path = r'C:\Users\bbissonette\Documents\Data Analytics Course\Acheivement 4\Instacart Basket Analysis'

In [5]:
# importing products data
df_prods = pd.read_csv(os.path.join(path,'02 Data', 'Original Data', 'products.csv'), index_col = False)

In [6]:
# importing orders wrangled data
df_ords = pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

## Consistency Checks

### Mixed-Type Data

In [7]:
# Create a dataframe
df_test = pd.DataFrame()

In [8]:
# Create a mixed type column
df_test['mix'] = ['a','b',1, True]

In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [10]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [11]:
# Changing column type from int to str
df_test['mix'] = df_test['mix'].astype('str')

### Missing Values

In [12]:
#  finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
# veiwing missing values in product mane col
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

### Addressing Missing Values

#### Create a new variable that acts like a flag based on the missing value

In [15]:
# finding the median of the col
df_prods.median()

C:\Users\bbissonette\AppData\Local\Temp\ipykernel_11736\2910585379.py:2: FutureWarning: The default value of numeric_only in DataFrame.median is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df_prods.median()


product_id       24845.0
aisle_id            69.0
department_id       13.0
prices               7.1
dtype: float64

In [16]:
# replacing missing values with median
df_prods['product_name'].fillna(24845, inplace=True)

In [17]:
# finding shape of prods df
df_prods.shape

(49693, 5)

In [18]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [19]:
df_prods_clean.shape

(49693, 5)

## Finding Duplicates

In [20]:
# defining and finding duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [21]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [22]:
df_prods_clean.shape

(49693, 5)

In [23]:
# defining a new dataframe with no duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [24]:
# Checking the shape of the new df to see if the dups were removed
df_prods_clean_no_dups.shape

(49688, 5)

# Task 4.5

### Step 2

In [25]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week    order_hour  days_since_prior_order  
count        3.421083e+06  3.421083e+06            3.214874e+06  
mean         2.776219e+00  1.345202e+01            1.111484e+01  
std          2.046829e+00  4.226088e+00            9.206737e+00  
min          0.000000e+00  0.000000e+00            0.000000e+00  
25%          1.000000e+00  1.000000e+01            4.000000e+00  
50%          3.000000e+00  1.300000e+01            7.000000e+00  
75%          5.000000e+00  1.600000e+01            1.500000e+01  
max          6.000000e+00  2.300000e+01            3.000000e+01

In [26]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49688.000000  49688.000000   49688.000000  49688.000000
mean   24844.500040     67.769582      11.728687      9.994254
std    14343.834402     38.316162       5.850410    453.542503
min        1.000000      1.000000       1.000000      1.000000
25%    12422.750000     35.000000       7.000000      4.100000
50%    24844.500000     69.000000      13.000000      7.100000
75%    37266.250000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

There arent any concerns in the ords data frame but the prods data frame has a incorrect value somewhere due to the max price being beyond reason.

In [27]:
# Creating a Subset to identify what product is priced at 99999
df_prods_clean_no_dups.loc[df_prods_clean_no_dups['prices'] == 99999.000000]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

So 2% Reduced Fat Milk had a typo when input into the system

### Checking for Mixed Values

In [30]:
# Check for mixed types in ords data frame
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
        print (col, ' mixed')
  else: print(col, ' consistent')

Unnamed: 0  consistent
order_id  consistent
user_id  consistent
eval_set  consistent
order_number  consistent
orders_day_of_week  consistent
order_hour  consistent
days_since_prior_order  consistent


All the columns have consistent data, so no need to fix anything

### Checking for missing values

In [31]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour                     0
days_since_prior_order    206209
dtype: int64

Many missing values in days since prior order

In [32]:
# Creating a Subset for the missing data
df_ords_missing_dspo = df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [33]:
# Viewing the new subset
df_ords_missing_dspo

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour  days_since_prior_order  
0                         2           8                     NaN  
11                        2          11                     NaN  
26                        1          14                     NaN  
39                        6          11                     NaN  
45                        3          12                     NaN  
...                     ...         ...                     ...  
3420930                   4          12                     NaN  
3420934                   3          18                     NaN  
3421002                   6          19                     NaN  
3421019                   1          15                     NaN  
3421069                   3          11                     NaN  

[206209 rows x 8 columns]

In [36]:
df_ords[df_ords['user_id'].isin([1,2,3])]

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0            0   2539329        1    prior             1                   2   
1            1   2398795        1    prior             2                   3   
2            2    473747        1    prior             3                   3   
3            3   2254736        1    prior             4                   4   
4            4    431534        1    prior             5                   4   
5            5   3367565        1    prior             6                   2   
6            6    550135        1    prior             7                   1   
7            7   3108588        1    prior             8                   1   
8            8   2295261        1    prior             9                   1   
9            9   2550362        1    prior            10                   4   
10          10   1187899        1    train            11                   4   
11          11   2168274        2    prior             1                   2   
12          12   1501582        2    prior             2                   5   
13          13   1901567        2    prior             3                   1   
14          14    738281        2    prior             4                   2   
15          15   1673511        2    prior             5                   3   
16          16   1199898        2    prior             6                   2   
17          17   3194192        2    prior             7                   2   
18          18    788338        2    prior             8                   1   
19          19   1718559        2    prior             9                   2   
20          20   1447487        2    prior            10                   1   
21          21   1402090        2    prior            11                   1   
22          22   3186735        2    prior            12                   1   
23          23   3268552        2    prior            13                   4   
24          24    839880        2    prior            14                   3   
25          25   1492625        2    train            15                   1   
26          26   1374495        3    prior             1                   1   
27          27    444309        3    prior             2                   3   
28          28   3002854        3    prior             3                   3   
29          29   2037211        3    prior             4                   2   
30          30   2710558        3    prior             5                   0   
31          31   1972919        3    prior             6                   0   
32          32   1839752        3    prior             7                   0   
33          33   3225766        3    prior             8                   0   
34          34   3160850        3    prior             9                   0   
35          35    676467        3    prior            10                   3   
36          36    521107        3    prior            11                   0   
37          37   1402502        3    prior            12                   1   
38          38   2774568        3     test            13                   5   

    order_hour  days_since_prior_order  
0            8                     NaN  
1            7                    15.0  
2           12                    21.0  
3            7                    29.0  
4           15                    28.0  
5            7                    19.0  
6            9                    20.0  
7           14                    14.0  
8           16                     0.0  
9            8                    30.0  
10           8                    14.0  
11          11                     NaN  
12          10                    10.0  
13          10                     3.0  
14          10                     8.0  
15          11                     8.0  
16           9                    13.0  
17          12                    14.0  
18          15          

The missing values are legit but they represent the first order from every customer. As you can see above the first through third customers first orders are all NaN in the days since prior order. The number of users in the system is 206209 so each customer has a first purchase which coincides with the number of NaN values.

### Addressing Missing Values

It doesnt make sense to impute a mean value or remove the rows of data. The information even though its identified as missing still provides value by indicating its the buyers first purchase. 

### Checking for Duplicates

In [37]:
# defining and making a duplicates subset
df_ords_dups = df_ords[df_ords.duplicated()]

In [38]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour, days_since_prior_order]
Index: []

In [39]:
df_ords_dups.size

0

No Duplicates found in the Orders Data Frame so no actions needed to remediate the duplicates.

In [40]:
# Exporting checked prods data frame
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [41]:
# Exporting Checked Ords Data Frame
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))